### ISSUE TITLE
Allow discarding out of bounds indexed draw calls

### ISSUE THREAD

**Issue by iraiter2** [gpuweb/gpuweb#320]
We believe that some implementations would benefit (when robust buffer access isn't available) from being able to discard indexed draws that have out of bounds values in the index buffer. Assuming we eventually have a debug mode, we could still alert the user that the draw call was discarded.

Here is one potential implementation we believe would benefit from being able to discard these calls: [Google doc](https://docs.google.com/document/d/1Vwlt6tN5-Xc5Es2aOKJSvM7H6ICEXMLf23M4QCKmS4U/)

**Comment by kainino0x** [comment]
PTAL, we'd like to see if we can come to an agreement next meeting.

**Comment by magcius** [comment]
Would this validation still be required if the indexes are uploaded on the CPU? It seems like a crazy thing to require for every draw call.

**Comment by jdashg** [comment]
This would exacerbate portability issues. What platforms do we expect to run on that do not have RBAB?

**Comment by magcius** [comment]
Vulkan's robustBufferAccess is not guaranteed to give you 0s when reading out of range, only to not crash.

**Comment by jdashg** [comment]
FWIW, in Firefox, we currently just cache `[start, end] => max_index` queries for index buffers, and invalidate the cache map on buffer updates.
Assuming most buffers are immutable and that indexed draw ranges repeat (i.e. they aren't completely random), this works quite well.

Firefox *used* to use a binary tree with wide leaves for tracking max_index, but it's hard to get good cache locality/behavior with this approach.

**Comment by jdashg** [comment]
> Vulkan's robustBufferAccess is not guaranteed to give you 0s when reading out of range, only to not crash.
Right, but it's perhaps attractive to just surface the RBA behavior to WebGPU, for better or worse, and rely on debug/development checks to point out when this happens.

**Comment by magcius** [comment]
I think the concern is security -- drivers have been known to leak other surface contents or uninitialized data from OOB reads.

**Comment by jdashg** [comment]
RBA is pretty constrained:
```
 Out-of-bounds buffer loads will return any of the following values:
▪ Values from anywhere within the memory range(s) bound to the buffer (possibly including bytes of memory past the end of the buffer, up to the end of the bound range).
▪ Zero values, or (0,0,0,x) vectors for vector reads where x is a valid value represented in the type of the vector components and may be any of:
▪ 0, 1, or the maximum representable positive integer value, for signed or unsigned integer components
▪ 0.0 or 1.0, for floating-point components
```

**Comment by kainino0x** [comment]
As far as I could tell, Metal has no robust buffer access behaviors, and is perfectly happy to GPU-reset if a draw is out of bounds, an index in an index buffer is out of bounds, or a compute shader reads/writes out of bounds of a storage buffer.

Maybe this has changed in Metal 3.

**Comment by kainino0x** [comment]
Hm, speaking of Metal 3, I can't find ANYTHING announcing it. But the [feature set tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) have changed (that's how I noticed) and there's [one mention](https://www.google.com/search?q=site%3Aapple.com+"Metal+3.0") of it in the documentation.

**Comment by jdashg** [comment]
@litherum What's the OOB access story on Metal?

**Comment by kvark** [comment]
> RBA is pretty constrained:
> 
> ```
>  Out-of-bounds buffer loads will return any of the following values:
> ▪ Values from anywhere within the memory range(s) bound to the buffer (possibly including bytes of memory past the end of the buffer, up to the end of the bound range).
> ▪ Zero values, or (0,0,0,x) vectors for vector reads where x is a valid value represented in the type of the vector components and may be any of:
> ▪ 0, 1, or the maximum representable positive integer value, for signed or unsigned integer components
> ▪ 0.0 or 1.0, for floating-point components
> ```

Wow, I wouldn't call "Values from anywhere within the memory range(s)" constrained :) This would to be a portability disaster if we allow such behavior at random.

**Comment by kainino0x** [comment]
Unfortunately since RBA is driver or hardware supported, we generally expect it to perform much better than anything we can do (such as running lg(n) compute shaders on index buffers).

**Comment by kainino0x** [comment]
But I think Jeff's point was that RBA (if implemented according to spec) is not a security risk. (Although there technically could be if there were a really weird implementation where the choice from that list depends on secret data.)

**Comment by magcius** [comment]
The goals of WebGPU are in the order: Security, Performance, Portability. We can have a debug mode that alerts the user if they're making invalid draw calls, but it might not make sense for it to be the default. If we do not need to validate draws for security reasons, performance should take priority over portability.

**Comment by kainino0x** [comment]
If a performance impact is very small compared with the portability impact, portability can still win. But in this case the performance impact is not negligible, even if we super-optimize (e.g. caching min/max for previously-used ranges of readonly index buffers, or something).

**Comment by kvark** [comment]
Just to have a comprehensive set of choices on the table, we should also consider doing manual vertex buffer fetches in the vertex shaders. If I'm not mistaken, this would still allow users to have index buffers, and the vertex cache in hardware would still be working. Therefore, there is a chance that performance will be sufficient, while we can strictly make it portable.

First thing that this would entail (that comes to mind) is that we'd want to move the vertex buffer definitions to the pipeline layout, which affects our discussions on:
  - removing the pipeline layout as an object (negative effect)
  - #145 (makes it trivial for Metal to support WebGPU without using indirect argument buffers)

**Comment by kainino0x** [comment]
I think this is what was mention here and discussed under "5" on [the investigation thread](https://github.com/gpuweb/gpuweb/issues/31#issuecomment-333889968).

I don't really know how this works (vertex pulling?). Is it available on the APIs where we would need it? (Metal, not sure about others)

And do we want to force implementations to use an implementation like that instead of another implementation, such as the segment tree?

**Comment by litherum** [comment]
In Metal, the OOB story is: don't do it, its undefined behavior.

**Comment by kvark** [comment]
To re-iterate: Metal *requires* us to handle out-of-bounds in the shader. Coincidentally, Metal also exposes the "pull" model for vertex buffers (the shader reads the data using VertexID).

D3D12 is able to return 0 and ignore writes out of bounds for buffers that are inside descriptor tables (as opposed to - inside the root signature directly). @RafaelCintron could you clarify if vertex buffers getting out of bounds is handled the same way?

Vulkan provides the "robustBufferAccess" feature, that (when enabled) can do whatever it wants with the assigned buffer memory, as long as it doesn't crash. It is also poisonous:
> If any buffer access is determined to be out of bounds, then any other access of the same type (load, store, or atomic) to the same buffer that accesses an address less than 16 bytes away from the out of bounds address may also be considered out of bounds

Basically, in Vulkan with RBA, if there is any access OOB, all bets about the assigned memory are off... Relying on RBA therefore is not a path that leads to a portable API.

**Comment by jdashg** [comment]
WebGL is pretty successfully portable and relies on RBA.

**Comment by iraiter2** [comment]
We made a doc summarizing the different viewpoints, feel free to suggest changes or comment: [doc link](https://docs.google.com/document/d/1UFwJLepiwvBBfUDn9_ImzKRXCvOKFFNpbdxJrex7u1o/)

**Comment by kenrussell** [comment]
Following up to comments from #311 : I think that discarding the draw call (no-op'ing it) should be one option available to implementations. Per discussion on the other issue, on platforms that don't support robust_buffer_access_behavior, the two most viable options for making indirect draw calls robust are either to no-op invalid draw calls, or transform *all* shaders to do programmable vertex pulling. The latter option is not viable in my opinion as it will cause large behavioral differences compared to how native applications work on these platforms.

**Comment by jdashg** [comment]
If you can detect invalidity for no-opping, can we then just rewrite the indirect buffers? That would preserve more correctness-portability between RBA and non-RBA systems.

Can we get RBA on Metal? It seems to be everywhere else!

**Comment by kenrussell** [comment]
@jdashg did you mean the index buffer or indirect buffer?

@iraiter2 's prototype does rewrite the indirect buffer; it copies it to a scratch buffer using a compute shader, validating each draw call described in the indirect buffer, and no-op'ing it if it's invalid. This approach seems to copy the minimal amount of data, as opposed to copying for example the index or vertex buffers.

@litherum @grorg would it be possible to submit a feature request for Metal that would allow enabling more robust behavior for a MTLDevice, at a performance cost? Not sure how this would best be done, i.e., exposing alternate devices which support robust access (would that work on iOS?), a run-time flag that can be enabled early in a MTLDevice's lifetime, or something else.

**Comment by jdashg** [comment]
Either index or indirect, really.

Is your goal to make OOB fetches fast? Because I think we should happily slow-path those.

**Comment by kainino0x** [comment]
Rewriting the indirect buffer (the 5 ints) won't help us make the indirect call valid; we would have to no-op. It's the contents of the index buffer which are a problem.

Are you alluding to doing index buffer rewrites (which would be very expensive), but only if the values in the index buffer would be invalid? We considered this at some point (probably should have written it down), but I think it requires a GPU-CPU roundtrip: a compute shader determines whether rewriting is necessary, we report the result back to the CPU, the CPU rewrites the draw call if necessary. I suppose on Metal it's probably possible to avoid the roundtrip by using argument buffers?

**Comment by iraiter2** [comment]
Test from the meeting notes:
> A modified version of Apple’s Hello Triangle from https://developer.apple.com/metal/sample-code/
500,000 triangles were rendered with position and color. One method used vertex attributes (push), and the other used a buffer to pull position / color based on vertex_id. FPS was measured using xcode.
IR: Results:
We ran the test on Nvidia / AMD / Intel and found no difference in FPS. This indicates that pulling works about just as well as pushing. However, we didn’t test mobile devices yet. Benchmarks on Nvidia with OpenGL from https://www.reddit.com/r/opengl/comments/6mj2uz/opengl_programmable_pulling_benchmark/ suggest it may be very slightly slower.

We ran the same test on:
iPad Air A1474 with 500k tris: 20 fps for both pull and push
iPhone 7+ with 1 million tris: 45 fps for both pull and push
iPhone xs max with 2 million tris: 30 fps for both pull and push
Looking at the frame time which had slightly more precision, there was still no difference.

**Comment by kvark** [comment]
About the approach of replacing all indexed calls by indirect indexed calls. Is this suggested as a general solution? It seem to me that doing this may not even be possible:
  - on Metal, "Indirect draw and dispatch arguments" are [not available](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) on MTLGPUFamilyApple1, MTLGPUFamilyApple2, and MTLGPUFamilyCommon1
  - on Vulkan, "drawIndirectFirstInstance" feature is [not available](https://vulkan.gpuinfo.org/listreports.php?feature=drawIndirectFirstInstance&option=not) on 1/9th of devices

**Comment by kainino0x** [comment]
I think you're right that technique wouldn't work everywhere.

**Comment by devshgraphicsprogramming** [comment]
I think just clamping the index values (either via programmable pulling or shadow copy of index buffer) could work better than no-op'ing.

Take worst case scenario, shadow copy, yes it will take more memory than your B-Tree or the segment tree, but trees are notoriously inefficient to construct on GPUs which are massively parallel, you either use atomics and painful synch or waste memory, also workload distribution is a problem.

Also lets not forget index buffers are 2 to 4 byte indices = 6 to 12 bytes per triangle
Whereas the most modest vertex type will be about 18 bytes (12 pos, 4 color, 4 uv packed), and average vertex re-use for normal meshes is 1.5, so that would make 24 bytes per triangle.

TL;DR vertex data storage cost dwarfs index buffer storage cost.

I don't really see why you'd need a segment tree for clamping, you already know the pipeline bound for the drawcall and hence the shader input buffer ranges bound, those are are segments... I don't see a need for a segment tree as ranges are already known. Am I missing something ?

**Comment by kainino0x** [comment]
The segment tree would be used to check the min/max for the actually-used range of the index buffer (determined by the setIndexBuffer offset and draw call indexCount). If the values in the that range of the buffer are out of bounds, we would no-op the call (not clamp the index values).

**Comment by devshgraphicsprogramming** [comment]
But then you'd have to query the range-tree for every draw call, there most likely will be interdependent sections where you can only check 1 drawcall ahead + fences that will prevent async/overlapped execution. Example: compute shader generates index buffer for indirect draw, then indirect draw happens immediately after, so there's only one place your validation CS can go.

If you don't have a batch of hundreds of drawcalls to validate then you're gonna seriously undersaturate the GPU, as no-oping a drawcall is basically a reduction to a bitwise mask (make drawcount 0).

GPU Reductions either require painfully slow AND or SET/EXCHANGE atomics (at least one per work-group) or multiple passes with memory barriers inbetween.


Programmable Vertex Pulling could be implemented with UTB/TBO on all GPUs (TBO reads in vx shader are not SSBO access), could be cached and have no memory overhead. Plus TBO implements most vertex formats as texel formats with automatic conversions ;)

**Comment by kainino0x** [comment]
I think it was agreed that vertex pulling is most likely the best approach when we don't have robust buffer access (which is only on Metal, where Idan [demonstrated](https://github.com/gpuweb/gpuweb/issues/320#issuecomment-502893619) that vertex pulling works well, and on unreliable Vulkan drivers). Thank you for the detailed analysis, it makes things more concrete.

Anyway, given the pushback on this proposal, I think we are willing to drop it until it comes up again.

**Comment by kainino0x** [comment]
Actually, I remembered something @kvark said [early on this thread](https://github.com/gpuweb/gpuweb/issues/320#issuecomment-499976916).

Doing vertex pulling (even inside the implementation) requires us to know the vertex buffer layout at pipeline layout creation. This looks like it won't affect Metal (where we don't use the pipeline layout until pipeline creation time anyway), but it does mean that doing a workaround for an unreliable Vulkan driver becomes more painful (have to delay pipeline layout creation until pipeline creation time).

Are there any more thoughts on that?

**Comment by devshgraphicsprogramming** [comment]
Regarding pipeline layout creation, I don't see why you'd need to delay pipeline layout creation until pipeline creation time. On Vulkan the pipeline includes the vertex buffer layout, so you can hotpatch the pipeline layout quite easily to remove vertex inputs and replace them with Uniform Texel Buffers.

The lovely thing about UTBs is that they count against the texture slot limit, and not against the SSBO limit.

I wouldn't try to over-optimize and try and put more than 2 vertex buffers in one UTB.

You have a choice of:
- One BufferView per VkBuffer used as UTB for a Vertex Buffer Input
- Multiple BufferView's with offsets per VkBuffer , this might be better because max UTB ranges bound could be as small as 64k

**Comment by kvark** [comment]
> On Vulkan the pipeline includes the vertex buffer layout, so you can hotpatch the pipeline layout quite easily to remove vertex inputs and replace them with Uniform Texel Buffers.

If we are doing this implicitly, what descriptor set do we put the vertex buffers in? Reserving a set for this sounds like an easy solution, but then on some systems we only have 4 sets maximum, and so we reducing it to 3 could be quite limiting for the user.

**Comment by kainino0x** [comment]
We resolved on the call today to allow the originally suggested behavior (allow discarding indexed draw calls which result in out of bounds reads).

**Comment by kainino0x** [comment]
(Pending input from @litherum who wasn't there.)
